### DISCUSSION TITLE
Prisma + Next.js challenges

### DISCUSSION THREAD

**Discussion by patjakubik** [prisma/prisma#4556] [Q&A] [answered]
I’m having a couple challenges that I wanted to share with you guys.

So, first - Next.js can be deployed to Vercel where it is backed by serverless function, or it can run as regular Node.js server on e.g. Heroku, right?

1. Does Prisma need to be implemented differently depending on how Next is deployed?
2. If so - does documentation assume deployment to Vercel or deployment as regular Node.js server?

By implementation, I mean:

1. Should Prisma Client be instantiated only once or every time when it is needed?
2. Should disconnect() be called?

In my project hosted on Heroku, I have it setup like here: https://github.com/prisma/blogr-nextjs-prisma/tree/final - however I’m still dealing with multiple connections to my database.

**Comment by nikolasburk** [comment, marked as answer]
Hey Patryk, thanks a lot for raising this!

> Next.js can be deployed to Vercel where it is backed by serverless function, or it can run as regular Node.js server on e.g. Heroku, right?

Yes, as far as I understand this is correct. 

> Does Prisma need to be implemented differently depending on how Next is deployed?

When deployed to Vercel, the API routes are deployed to serverless functions. In general, it is [recommended](https://www.prisma.io/docs/concepts/components/prisma-client/deployment#recommended-connection-limit) to set the `connection_limit` argument to `1` when deploying apps that use Prisma Client to serverless functions:

```
postgresql://USER:PASSWORD@HOST:PORT/DATABASE?connection_limit=1
```

> In my project hosted on Heroku, I have it setup like here: https://github.com/prisma/blogr-nextjs-prisma/tree/final - however I’m still dealing with multiple connections to my database.

That's a great point, this resource actually needs to be updated to use a connection limit of `1` as well! We'll update it as soon as possible.

> Should Prisma Client be instantiated only once or every time when it is needed?

As far as I'm aware, it's generally recommended to create only a single instance of `PrismaClient`.

**Reply by patjakubik** [reply]
@nikolasburk 
Thanks for your answer Nikolas.

Okay, so I added `connection_limit=1` argument, opened my app in two tabs and I can see that connections went up to two in a period of one minute. Is that expected? My database has currently a limit of 20 connections, I'm worried that if more than 20 users use my app simultaneously, they will start running into issues. Is this normal? Or I'm missing something?

<img width="890" alt="CleanShot 2020-12-08 at 17 38 49@2x" src="https://user-images.githubusercontent.com/37963339/101513243-4158d200-397c-11eb-9759-1aea1ab1624a.png">

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
